File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let answer = 0;
    
    // 오른쪽(동생)의 토핑 종류별 개수와 총 종류 수
    const rightMap = new Map();
    let rightCount = 0;
    
    for (const t of topping) {
        if (!rightMap.has(t)) {
            rightMap.set(t, 0);
            rightCount++;
        }
        rightMap.set(t, rightMap.get(t) + 1);
    }
    
    // 왼쪽(철수)의 토핑 종류를 담을 셋(Set)
    const leftSet = new Set();
    
    // 롤케이크를 왼쪽/오른쪽으로 나누며 확인
    for (const t of topping) {
        // 왼쪽으로 토핑 추가
        leftSet.add(t);
        
        // 오른쪽에서 토핑 제거
        rightMap.set(t, rightMap.get(t) - 1);
        
        // 오른쪽 토핑 개수가 0이 되면 종류 수 감소
        if (rightMap.get(t) === 0) {
            rightCount--;
        }
        
        // 철수와 동생의 토핑 종류 수가 같다면 방법의 수 1 추가
        if (leftSet.size === rightCount) {
            answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        
        // 1. 토핑의 종류는 1~10000 제한이므로, 개수를 셀 배열 선언
        int[] left = new int[10001]; // 철수
        int[] right = new int[10001]; // 동생
        
        int leftCount = 0;
        int rightCount = 0;
        
        // 2. 초기 세팅: 모든 토핑을 오른쪽(동생)에 몰아주기
        for (int t : topping) {
            if (right[t] == 0) rightCount++;
            right[t]++;
        }
        
        // 3. 포인터를 이동하며 왼쪽(철수)으로 토핑 하나씩 옮기기
        for (int t : topping) {
            // 왼쪽(철수)에 추가
            if (left[t] == 0) leftCount++;
            left[t]++;
            
            // 오른쪽(동생)에서 제거
            right[t]--;
            if (right[t] == 0) rightCount--;
            
            // 4. 두 조각의 토핑 종류 수가 같다면 정답 카운트 추가
            if (leftCount == rightCount) {
                answer++;
            }
        }
        
        return answer;
    }
}
